[Routing] Deprecate `RouteCollection` overriding its routes' properties

MatTheCat opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 7.4
| Bug fix?      | no
| New feature?  | no
| Deprecations? | yes
| Issues        | N/A
| License       | MIT

In Symfony, route collections’ methods, schemes, host, defaults, requirements and condition end up overriding their routes’ which is counter-intuitive: it means e.g. every route in a collection whose method is `POST` will have the same, even if they were configured through the `Route` attribute with a `GET` method.

Even if doing this has currently no effect and is probably a mistake, changing this behavior would be a BC break so this PR starts by deprecating it.

Then, https://github.com/symfony/symfony/issues/53171 can be fixed in 8.0 by changing the conditions and moving the call to routes’ setter into it, e.g.

```diff
foreach ($this->routes as $route) {
-    if ('' !== $route->getHost() && $route->getHost() !== $pattern) {
-        trigger_deprecation('symfony/routing', '7.3', 'Overriding a route\'s host with its group\'s is deprecated, you should remove it from the route.');
+    if ('' === $route->getHost()) {
+        $route->setHost($pattern);
    }

-    $route->setHost($pattern);
}
```

I’ll be waiting for an approval before updating loaders tests because it looks like it’ll be tedious.

MatTheCat committed:
Deprecate `RouteCollection` overriding its routes' properties

ro0NL commented:
im wondering if we should rename the methods to `setDefaultSomething` instead then, to convey it does not overwrite

MatTheCat commented:
In my mind the collections’ properties already are the default for their routes’ so I’m not sure.

alamirault committed:
Replace get_class() calls by ::class

nicolas-grekas committed:
minor #60581 Replace `get_class()` calls by `::class` (alamirault)

This PR was merged into the 7.4 branch.

Discussion
----------

Replace `get_class()` calls by `::class`

| Q             | A
| ------------- | ---
| Branch?       | 7.4
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Issues        | Fix #... <!-- prefix each issue number with "Fix #", no need to create an issue if none exists, explain below instead -->
| License       | MIT

Replace `get_class()` by `::class`
It was already done in past in https://github.com/symfony/symfony/pull/47401

Commits
-------

e0a602bfc57 Replace get_class() calls by ::class
